### DISCUSSION TITLE
Analyzing system-level dependencies for Postgres connectivity

### DISCUSSION THREAD

**Discussion by aaronsteers** [meltano/meltano#6617] [General]
In office hours, we discussed that our new slimmer and more secure docker images do not have full support for Postgres connectivity in some cases.

I did a quick scan of repos and all of these are using psycopg2-binary:
- [Meltano itself](https://github.com/meltano/meltano/blob/main/pyproject.toml#L52)
- [pipelinewise-target-postgres](https://github.com/transferwise/pipelinewise-target-postgres/blob/master/setup.py#L22) (the Hub default)
- [pipelinewise-tap-postgres](https://github.com/transferwise/pipelinewise-tap-postgres/blob/master/setup.py#L21) (the Hub default)
- [MeltanoLabs/target-postgres](https://github.com/MeltanoLabs/target-postgres/blob/main/pyproject.toml#L16) (still WIP with @visch)
- [MeltanoLabs/tap-postgres](https://github.com/MeltanoLabs/tap-postgres/blob/main/pyproject.toml#L16)

Opening this discussion as follow-on to the [Slack conversation here](https://meltano.slack.com/archives/CKHP6G5V4/p1660322913798549?thread_ts=1660316207.458259&cid=CKHP6G5V4).

Let's discuss+document below what is required and what workaround might exist.

**Comment by aaronsteers** [comment]
Looks like this was expected to have been resolved in a prior Meltano release:

- [GitLab !1676 - Require `psycopg2-binary` instead of `psycopg2` so that build dependency `pg_config` doesn't need to be present on system](https://gitlab.com/meltano/meltano/-/merge_requests/1676)

**Comment by aaronsteers** [comment]
The `datamill-co/target-postgres` plugin one does not use the `-binary` Postgres package:

https://github.com/datamill-co/target-postgres/blob/master/setup.py#L23

**Reply by WillDaSilva** [reply]
I'll open a PR for their consideration that will update the dep to the `-binary` version.

**Reply by WillDaSilva** [reply]
While I was writing the description for the PR, I came across https://www.psycopg.org/docs/install.html#psycopg-vs-psycopg-binary. It states very strongly that packages should not depend on `psycopg2-binary`:

> The psycopg2-binary package is meant for beginners to start playing with Python and PostgreSQL without the need to meet the build requirements.
> 
> If you are the maintainer of a published package depending on psycopg2 you shouldn’t use psycopg2-binary as a module dependency. For production use you are advised to use the source distribution.
> 
> The binary packages come with their own versions of a few C libraries, among which libpq and libssl, which will be used regardless of other libraries available on the client: upgrading the system libraries will not upgrade the libraries used by psycopg2. Please build psycopg2 from source if you want to maintain binary upgradeability.
> 
> Warning: The psycopg2 wheel package comes packaged, among the others, with its own libssl binary. This may create conflicts with other extension modules binding with libssl as well, for instance with the Python [ssl](https://docs.python.org/3/library/ssl.html#module-ssl) module: in some cases, under concurrency, the interaction between the two libraries may result in a segfault. In case of doubts you are advised to use a package built from source.

In light of this, it may be difficult to convince `datamill-co/target-postgres` to switch to `postgres2-binary`. More than that, it may even be inadvisable for them to do so, and for us to stick with it.

**Reply by aaronsteers** [reply]
@WillDaSilva - Thanks for noting this. We've run into the disclaimer previously several times, and in practice, it's just not a strong enough argument to sacrifice portability and specifically, not strong enough argument to require OS-level dependency installations when otherwise none would be needed.

My observation is that those warnings are likely overblown. For instance, I've never seen or heard of a conflict occuring in the wild such as the mentioned "segfault" issue. 

Each maintainer certainly is entitled to make that decision themselves, but maintainers overall (including Meltano) generally have needed to migrate anyway in order to provide sufficient interoperability and portability.

In the case of `target-postgres`, we do also have the Pipelinewise variant (our current default on the Hub) which we can recommend to the user if they do not want to manage the pre-install on a custom image.

**Reply by WillDaSilva** [reply]
PR opened: https://github.com/datamill-co/target-postgres/pull/219

**Reply by visch** [reply]
Just found out why this feels like a such a moving target as this issue also happens with `transferwise-target-postgres` :O the reason is that they have an outdated version of `psycopg2-binary`  which doesn't have wheels for `Python 3.9.*` , note that I couldn't actually get the variant working on Python 3.8 I had to use the Meltano python 3.7 version by updating my Docker file to be

```
ARG MELTANO_IMAGE=meltano/meltano:latest-python3.7
FROM $MELTANO_IMAGE
```

Another fix is to point to this version of target-postgers
* https://github.com/transferwise/pipelinewise-target-postgres/pull/65
I also confirmed that this will work as well

So this means that just having `pipelinewise-binary` isn't enough we also need to have the latest version (Or at least one with wheels for Python 3.8 / 3.9 / etc) 

Side bar:

- Locally it looks like podman does something with local executables that throws everything off (Leading me down a rabbit hole, note to self and others if you install libpq somehow podman's containers can use these binaries, no idea how or what's happening here as I don't understand enough right now to know why this happens without dumping a bunch of time)

**Comment by aaronsteers** [comment]
There's a version of `target-redshift` which also requires this:  https://github.com/meltano/meltano/issues/2019#issuecomment-1213637755

Discussed here: https://github.com/meltano/meltano/issues/2019#issuecomment-1215321116

Link to source code here: https://github.com/datadotworld/tap-redshift/blob/master/setup.py#L54

**Reply by aaronsteers** [reply]
@DouweM, @visch, @WillDaSilva, @tayloramurphy 

### Proposed guidance

Knowing that the postgres libraries do expand our base security footprint (at least slightly), I'm inclined to add the following as guidance but _not_ add back the base install at this time:

1. Provide guidance for how users can add the dependency to their own images as needed.
2. Provide guidance for when users can switch to another variant that doesn't require the OS-level install.

### Proposed image support plan

Regarding the supported images, I don't want to add this back in and then have to remove again in the future. I can see two paths that seem long-term viable:

1. We can provide a _temporary_ image (or image set) that has the postgres installs already installed, with a note that a future version of Meltano will discontinue it, and users are encouraged to build their own image off of `meltano/meltano`. The idea would be to keep the base image clean as possible while still giving a pragmatic workaround to those affected. 
2. Somewhat reverse of this, perhaps better for long-term, would be to go ahead and install the postgres image into our main `meltano` images, but introduce a `meltano-slim` set that intentionally does not have those extra libraries. A benefit of this approach is that it balances interop/ease-of-use in the "default" package that folks will use locally and when evaluating meltano, while also having long term support for whatever we think the "most slim+secure" image would be (addressing the "Day 2" challenge of deploying Meltano in a hardened/secured manner).

### Assumptions

This is based on the understanding that we only have a few taps in the ecosystem which still require the non-binary Postgres library, and based on the fact that users have a fairly good workaround which is to install the OS-level dependencies on a derived Docker image (same as they'd do for any other OS-level dependencies needed by other plugins).

**Reply by tayloramurphy** [reply]
@aaronsteers without having spent a ton of time thinking about it, it seems 2 is preferable. It gives a stable default experience but offers a path where users can have more control.

**Comment by aaronsteers** [comment]
We have new news here as of Pyscopg 3:

- https://www.psycopg.org/psycopg3/docs/basic/install.html#binary-installation

----------

> <img width="763" alt="image" src="https://user-images.githubusercontent.com/18150651/184977439-234f86e4-068b-4e46-bf56-296cd2373129.png">

----------

This apparently won't work great on M1 Macs so it might not be immediately actionable. That said, it does seem to pave the way for a lower-friction future.

---------- 

Looks like most changes from `pyscopg2` and `psycopg v3.x` are intended to be backwards compatible:

https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html

> 
<img width="784" alt="image" src="https://user-images.githubusercontent.com/18150651/184978346-79ccd8e7-c753-4900-b3f3-8b7bdeffd5bc.png">

**Reply by WillDaSilva** [reply]
This is fantastic news! Even if other components don't provide the `binary` extra, we can, and it'll be picked up automatically.
